lnd.go

```diff
@@ -251,33 +251,79 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
 		"minutes...")
 
 	startOpenTime := time.Now()
+
 	databaseBackends, err := cfg.DB.GetBackends(
 		cfg.localDatabaseDir(), cfg.networkName(),
 	)
-	if err != nil {
-		ltndLog.Error(err)
-		return err
-	}
 
-	// Open the channeldb, which is dedicated to storing channel, and
-	// network related metadata.
-	chanDB, err := channeldb.CreateWithBackend(
-		databaseBackends.LocalDB,
-		channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
-		channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
-		channeldb.OptionSetSyncFreelist(cfg.SyncFreelist),
-		channeldb.OptionDryRunMigration(cfg.DryRunMigration),
-	)
-	switch {
-	case err == channeldb.ErrDryRunMigrationOK:
-		ltndLog.Infof("%v, exiting", err)
-		return nil
+	// If the remoteDB is nil, then we'll just open a local DB as normal,
+	// having the remote and local pointer be the exact same instance.
+	var localChanDB, remoteChanDB *channeldb.DB
+	if remoteChanDB == nil {
+		// Open the channeldb, which is dedicated to storing channel, and
+		// network related metadata.
+		localChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.LocalDB,
+			channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+			channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+			channeldb.OptionSetSyncFreelist(cfg.SyncFreelist),
+		)
+		switch {
+		case err == channeldb.ErrDryRunMigrationOK:
+			ltndLog.Infof("%v, exiting", err)
+			return nil
 
-	case err != nil:
-		ltndLog.Errorf("Unable to open channeldb: %v", err)
-		return err
+		case err != nil:
+			err := fmt.Errorf("Unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
+			return err
+		}
+		defer localChanDB.Close()
+
+		remoteChanDB = localChanDB
+	} else {
+		ltndLog.Infof("Database replication detected! Creating " +
+			"local and remote channeldb instances")
+
+		// Otherwise, we'll open two instances, one for the state we only
+		// need locally, and the other for things we want to ensure are
+		// replicated.
+		localChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.LocalDB,
+			channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+			channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+			channeldb.OptionSetSyncFreelist(cfg.SyncFreelist),
+		)
+		switch {
+		case err == channeldb.ErrDryRunMigrationOK:
+			ltndLog.Infof("%v, exiting", err)
+			return nil
+
+		case err != nil:
+			err := fmt.Errorf("Unable to open local channeldb: %v", err)
+			ltndLog.Error(err)
+			return err
+		}
+		defer localChanDB.Close()
+
+		remoteChanDB, err = channeldb.CreateWithBackend(
+			databaseBackends.RemoteDB,
+			channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+			channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+			channeldb.OptionSetSyncFreelist(cfg.SyncFreelist),
+		)
+		switch {
+		case err == channeldb.ErrDryRunMigrationOK:
+			ltndLog.Infof("%v, exiting", err)
```

Comment:
nit: Should there be a difference in behavior between the two cases? Right now both log and exit, so you could consider merging them.

Reply:
So they're both cases of running a database migration in the dry run mode. However the way it is now, we'll run it for the local then bail out all together which wouldn't be the correct behavior in this case. Modified to ensure both execute properly if this is indeed a dry run mode.